### PR TITLE
Expand `FallbackImage` to include a `GpuImage` for each possible `TextureViewDimension`

### LINKED ISSUE #6920: `FallbackImage` only supports `TextureDimension::D2`
## Bevy version

0.9

## What you did

I tried to use the `AsBindGroup` derive for a custom material that uses optional array textures:

```rust
#[derive(AsBindGroup, Reflect, FromReflect, Debug, Clone, TypeUuid)]
#[uuid = "2f3d7f74-4bf7-4f32-98cd-858edafa5ca2"]
#[bind_group_data(MyMaterialKey)]
#[uniform(0, MyMaterialUniform)]
#[reflect(Default, Debug)]
pub struct MyMaterial {
    pub base_color: Color,

    #[texture(1, dimension = "2d_array")]
    #[sampler(2)]
    pub base_color_texture: Option<Handle<Image>>,

   ...
}
```

## What went wrong

The application crashed while trying to create the bind group. See full crash output below. The reason was:

```
wgpu error: Validation Error

Caused by:
    In Device::create_bind_group
    texture binding 1 expects dimension = D2Array, but given a view with dimension = D2
```

### Theory of Root Cause

As the title suggests, I believe the root cause is that `FallbackImage` is only ever created with `TextureDimension::D2`. So when I provide `None` for the optional image, the fallback has a mismatch with the expected texture binding layout.

### stdout/stderr + backtrace

```
2022-12-11T23:41:31.770452Z  INFO winit::platform_impl::platform::x11::window: Guessed window scale factor: 1.25    
2022-12-11T23:41:33.156370Z  INFO bevy_render::renderer: AdapterInfo { name: "NVIDIA GeForce RTX 3070 Laptop GPU", vendor: 4318, device: 9373, device_type: DiscreteGpu, driver: "NVIDIA", driver_info: "510.85.02", backend: Vulkan }
2022-12-11T23:41:33.704555Z ERROR wgpu::backend::direct: Handling wgpu errors as fatal by default    
thread 'Compute Task Pool (2)' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_bind_group
    texture binding 1 expects dimension = D2Array, but given a view with dimension = D2

', /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.14.2/src/backend/direct.rs:2403:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   2: wgpu::backend::direct::default_error_handler
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.14.2/src/backend/direct.rs:2403:5
   3: core::ops::function::Fn::call
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:77:5
   4: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1954:9
   5: wgpu::backend::direct::ErrorSinkRaw::handle_error
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.14.2/src/backend/direct.rs:2389:17
   6: wgpu::backend::direct::Context::handle_error
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.14.2/src/backend/direct.rs:254:9
   7: <wgpu::backend::direct::Context as wgpu::Context>::device_create_bind_group
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.14.2/src/backend/direct.rs:1298:13
   8: wgpu::Device::create_bind_group
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.14.2/src/lib.rs:2170:17
   9: bevy_render::renderer::render_device::RenderDevice::create_bind_group
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_render-0.9.1/src/renderer/render_device.rs:75:31
  10: <bevy_triplanar_splatting::TriplanarMaterial as bevy_render::render_resource::bind_group::AsBindGroup>::as_bind_group
             at ./src/lib.rs:22:10
  11: bevy_pbr::material::prepare_material
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_pbr-0.9.1/src/material.rs:590:20
  12: bevy_pbr::material::prepare_materials
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_pbr-0.9.1/src/material.rs:566:15
  13: core::ops::function::FnMut::call_mut
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:164:5
  14: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:294:13
  15: <Func as bevy_ecs::system::function_system::SystemParamFunction<(),Out,(F0,F1,F2,F3,F4,F5,F6),()>>::run::call_inner
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.9.1/src/system/function_system.rs:579:21
  16: <Func as bevy_ecs::system::function_system::SystemParamFunction<(),Out,(F0,F1,F2,F3,F4,F5,F6),()>>::run
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.9.1/src/system/function_system.rs:582:17
  17: <bevy_ecs::system::function_system::FunctionSystem<In,Out,Param,Marker,F> as bevy_ecs::system::system::System>::run_unsafe
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.9.1/src/system/function_system.rs:409:19
  18: bevy_ecs::schedule::executor_parallel::ParallelExecutor::prepare_systems::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.9.1/src/schedule/executor_parallel.rs:218:26
  19: bevy_ecs::schedule::executor_parallel::ParallelExecutor::prepare_systems::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.9.1/src/schedule/executor_parallel.rs:258:21
  20: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  21: async_executor::Executor::spawn::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/async-executor-1.5.0/src/lib.rs:139:19
  22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  23: async_task::raw::RawTask<F,T,S>::run
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/async-task-4.3.0/src/raw.rs:511:20
  24: async_task::runnable::Runnable::run
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/async-task-4.3.0/src/runnable.rs:309:18
  25: async_executor::Executor::run::{{closure}}::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/async-executor-1.5.0/src/lib.rs:230:21
  26: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  27: <futures_lite::future::Or<F1,F2> as core::future::future::Future>::poll
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.12.0/src/future.rs:529:33
  28: async_executor::Executor::run::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/async-executor-1.5.0/src/lib.rs:237:31
  29: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  30: futures_lite::future::block_on::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.12.0/src/future.rs:89:27
  31: std::thread::local::LocalKey<T>::try_with
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/thread/local.rs:445:16
  32: std::thread::local::LocalKey<T>::with
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/thread/local.rs:421:9
  33: futures_lite::future::block_on
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.12.0/src/future.rs:79:5
  34: bevy_tasks::task_pool::TaskPool::new_internal::{{closure}}::{{closure}}::{{closure}}::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_tasks-0.9.1/src/task_pool.rs:128:37
  35: std::panicking::try::do_call
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:492:40
  36: __rust_try
  37: std::panicking::try
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:456:19
  38: std::panic::catch_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panic.rs:137:14
  39: bevy_tasks::task_pool::TaskPool::new_internal::{{closure}}::{{closure}}::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_tasks-0.9.1/src/task_pool.rs:122:43
  40: std::thread::local::LocalKey<T>::try_with
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/thread/local.rs:445:16
  41: std::thread::local::LocalKey<T>::with
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/thread/local.rs:421:9
  42: bevy_tasks::task_pool::TaskPool::new_internal::{{closure}}::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_tasks-0.9.1/src/task_pool.rs:120:25
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_tasks-0.9.1/src/task_pool.rs:273:45
stack backtrace:
   0: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:48:5
   3: core::option::Option<T>::unwrap
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/option.rs:775:21
   4: bevy_tasks::task_pool::TaskPool::scope::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_tasks-0.9.1/src/task_pool.rs:273:34
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
   6: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/future.rs:124:9
   7: <&mut F as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/future.rs:112:9
   8: <futures_lite::future::PollOnce<F> as core::future::future::Future>::poll
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.12.0/src/future.rs:194:15
   9: futures_lite::future::block_on::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.12.0/src/future.rs:89:27
  10: std::thread::local::LocalKey<T>::try_with
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/thread/local.rs:445:16
  11: std::thread::local::LocalKey<T>::with
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/thread/local.rs:421:9
  12: futures_lite::future::block_on
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.12.0/src/future.rs:79:5
  13: bevy_tasks::task_pool::TaskPool::scope
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_tasks-0.9.1/src/task_pool.rs:283:39
  14: <bevy_ecs::schedule::executor_parallel::ParallelExecutor as bevy_ecs::schedule::executor::ParallelSystemExecutor>::run_systems
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.9.1/src/schedule/executor_parallel.rs:127:9
  15: <bevy_ecs::schedule::stage::SystemStage as bevy_ecs::schedule::stage::Stage>::run
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_ecs-0.9.1/src/schedule/stage.rs:802:17
  16: <bevy_render::RenderPlugin as bevy_app::plugin::Plugin>::build::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_render-0.9.1/src/lib.rs:261:21
  17: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1954:9
  18: bevy_app::app::App::update
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_app-0.9.1/src/app.rs:154:13
  19: bevy_winit::winit_runner_with::{{closure}}
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_winit-0.9.1/src/lib.rs:600:21
  20: winit::platform_impl::platform::sticky_exit_callback
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.27.5/src/platform_impl/linux/mod.rs:849:9
  21: winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.27.5/src/platform_impl/linux/x11/mod.rs:363:17
  22: winit::platform_impl::platform::x11::EventLoop<T>::run_return
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.27.5/src/platform_impl/linux/x11/mod.rs:448:31
  23: winit::platform_impl::platform::x11::EventLoop<T>::run
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.27.5/src/platform_impl/linux/x11/mod.rs:503:25
  24: winit::platform_impl::platform::EventLoop<T>::run
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.27.5/src/platform_impl/linux/mod.rs:755:56
  25: winit::event_loop::EventLoop<T>::run
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.27.5/src/event_loop.rs:278:9
  26: bevy_winit::run
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_winit-0.9.1/src/lib.rs:263:5
  27: bevy_winit::winit_runner_with
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_winit-0.9.1/src/lib.rs:645:9
  28: bevy_winit::winit_runner
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_winit-0.9.1/src/lib.rs:303:5
  29: core::ops::function::Fn::call
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:77:5
  30: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1954:9
  31: bevy_app::app::App::run
             at /home/duncan/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_app-0.9.1/src/app.rs:168:9
  32: render::main
             at ./examples/render.rs:6:5
  33: core::ops::function::FnOnce::call_once
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:248:5
```

### PR DESCRIPTION
**Pull request by bonsairobo**
# Objective

Fixes #6920 

## Solution

From the issue discussion:

> From looking at the `AsBindGroup` derive macro implementation, the fallback image's `TextureView` is used when the binding's `Option<Handle<Image>>` is `None`. Because this relies on already having a view that matches the desired binding dimensions, I think the solution will require creating a separate `GpuImage` for each possible `TextureViewDimension`.

---

## Changelog

Users can now rely on `FallbackImage` to work with a texture binding of any dimension.

### CODE REVIEW

**Comment by bonsairobo** [comment]
Currently a draft b/c I haven't tested that this actually fixes the panic yet. I need to get my downstream crate (and it's dependencies that depend on `bevy`) building based on these changes first.

**Comment by bonsairobo** [comment]
Should I be putting the integration test into an example? Is that the right pattern?

**Comment by bonsairobo** [comment]
The CI failures seem unrelated to my changes.

**Review by IceSentry — APPROVED** [review]

**Comment by IceSentry** [comment]
Would be nice if the example was able to run as an actual test case instead of being an example, but this is out of scope of this PR since we don't have the infrastructure for this right now.

File: examples/shader/fallback_image.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+use bevy::{
```
Comment (alice-i-cecile) [inline examples/shader/fallback_image.rs:1]: This needs module-level doc strings explaining the purpose of the example.

File: crates/bevy_render/src/texture/fallback_image.rs [hunk 2]
```diff
@@ -20,8 +20,15 @@ use crate::{
 ///
 /// Defaults to a 1x1 fully opaque white texture, (1.0, 1.0, 1.0, 1.0) which makes multiplying
 /// it with other colors a no-op.
-#[derive(Resource, Deref)]
-pub struct FallbackImage(GpuImage);
+#[derive(Resource)]
+pub struct FallbackImage {
+    pub d1: GpuImage,
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/texture/fallback_image.rs:25]: Doc strings for all of these fields please: they're quite challenging to follow.

File: examples/shader/fallback_image.rs (hunk 1 above)
Comment (bonsairobo) [reply examples/shader/fallback_image.rs:1]: I'll just move the docs I wrote for the `main` function. I just copied from another example.

**Comment by bonsairobo** [comment]
Should be good to go.

**Comment by bonsairobo** [comment]
> Much clearer, thanks :) I still don't have quite enough expertise to review this directly though, so I'm going to hold off on an approval.

Ok. We have an approval from Charles. Is that good enough or what happens next?

**Comment by IceSentry** [comment]
It still needs a second approval, there's been a tiny bit of pushback on discord. I wouldn't say it's a blocker, but the gist of it is that this solution doesn't scale well to support the other attributes that aren't currently supported like the texture format for example.

Also, I don't mind my real name being here, but not everyone knows that it's me.

**Review by JMS55 — APPROVED** [review]
Looks good to me. I'm not really concerned about combinatorial explosion due to number of parameters. Off the top of my head, I can't see why a user would care about the texture format. Binding it and using it in a shader is the same either way. The only thing that matters is: what value you get out (black/white), texture dimensions, and what format you read values in from the shader (float, uint, etc. Anything but float would be very rare though). In fact, we could use something like r32float/uint (or r8 on native) to save a very tiny amount of memory. 

Tldr; I don't see a current use case for larger amounts of fallbacks beyond what we have now.

**Comment by IceSentry** [comment]
I know that @DGriffin91 needed this for deferred shading which is why it was mentioned. It may not have been texture format, but it was definitely because of some other parameter not being handled.

**Comment by DGriffin91** [comment]
@IceSentry yes, texture format and sample count.

**Comment by bonsairobo** [comment]
@DGriffin91 @IceSentry If we want to support all possible `TextureFormat`s, then we probably don't want to have a field for each combination. We could use a `HashMap<(TextureViewDimension, TextureFormat), GpuImage>` and lazily create the images. This could happen in a follow-up issue/PR.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I've run the example locally and while it doesn't panic I'm just getting a grey screen. Is that intentional?

Either way, please add a description to the module comment at the top of the example explaining what "correct" behavior should look like.

**Comment by bonsairobo** [comment]
> I've run the example locally and while it doesn't panic I'm just getting a grey screen. Is that intentional?

Yea the shader doesn't do anything, it's just testing that it doesn't crash when attempting to bind the images.

> Either way, please add a description to the module comment at the top of the example explaining what "correct" behavior should look like.

Will do.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Much better, thanks!
